Remove the last remaining synchronize-via-sleep occurrences

arichardson opened the pull request:
The total runtime is now 4-5 seconds instead of almost 60 before the
changes to remove sleep().

arichardson committed:
Use volatile sig_atomic_t for signal handler variables

I was trying to update the Pdkill test to no longer use sleep() and noticed
that the read of the had_signal map was being optimized out by the compiler.
Writing to a std::map inside a signal handler is not safe, so replace it
with a sig_atomic_t array.

arichardson committed:
Avoid sleep() synchronization for Pdkill and PdkillSignal

As with the other tests use pipe()+read()/write() instead.

arichardson committed:
Remove the last sleep()-based synchronization from procdesc.cc

daviddrysdale commented on procdesc.cc:
```diff
@@ -245,19 +249,36 @@ static void *SubThreadMain(void *) {
     pthread_t child_subthread;
-    EXPECT_OK(pthread_create(&child_subthread, NULL, SubThreadMain, NULL));
+    // Wait for the subthread startup using atomics and p.
+    // Note: using the __sync compiler intrinsics to avoid depending on C++11.
```
Why does this need to use a new and different synchronization primitive?  Wouldn't it just be possible to create another pipe and pass one of the file descriptors as an argument to the new thread, then sync with `SEND_INT_MESSAGE` ?

arichardson replied:
Yes good point, not sure why I thought I shouldn't be using pipes here.

daviddrysdale commented on procdesc.cc:
```diff
@@ -853,51 +874,80 @@ TEST_F(PipePdfork, WildcardWait) {
-    had_signal.clear();
+    // Child: set a SIGINT handler, notify the parent and sleep.
+    close(pipefds[0]);
+    clear_had_signals();
```
Might be easier if `clear_had_signals()` also reset `last_signal` ?

daviddrysdale commented on procdesc.cc:
```diff
@@ -853,51 +874,80 @@ TEST_F(PipePdfork, WildcardWait) {
+    if (verbose) fprintf(stderr, "[%d] child started\n", getpid_());
+    SEND_INT_MESSAGE(pipefds[1], MSG_CHILD_STARTED);
+    // Child: wait for SIGINT for up to 50 * 100ms seconds.
+    for (volatile int i = 0; i < 50; i++) {
```
Why does the `int` need to be `volatile`?

arichardson replied:
That was a workaround when I was debugging why had_signal[SIGINT] was not being read (fixed by using volatile sig_atomic_t instead of std::map<>), so is not longer needed.

daviddrysdale commented on procdesc.cc:
```diff
@@ -853,51 +874,80 @@ TEST_F(PipePdfork, WildcardWait) {
+    // Child: wait for SIGINT for up to 50 * 100ms seconds.
+    for (volatile int i = 0; i < 50; i++) {
+      if (had_signal[SIGINT]) {
+        if (verbose)
```
nit: I had a convention of only skipping the `{` `}` for tracing if it would fit on a single line.

daviddrysdale commented on procdesc.cc:
```diff
@@ -853,51 +874,80 @@ TEST_F(PipePdfork, WildcardWait) {
+      }
+      usleep(100000);
+    }
+    if (verbose)
```
`{` `}`

daviddrysdale commented on procdesc.cc:
```diff
@@ -853,51 +874,80 @@ TEST_F(PipePdfork, WildcardWait) {
+      if (had_signal[SIGINT]) {
+        if (verbose)
+          fprintf(stderr, "[%d] child saw SIGINT after %d iterations\n",
+                  getpid_(), (int)had_signal[SIGINT]);
```
wrong final argument?

daviddrysdale commented on procdesc.cc:
```diff
@@ -853,51 +874,80 @@ TEST_F(PipePdfork, WildcardWait) {
-    exit(left == 0);
+    if (verbose) fprintf(stderr, "[%d] child started\n", getpid_());
+    SEND_INT_MESSAGE(pipefds[1], MSG_CHILD_STARTED);
+    // Child: wait for SIGINT for up to 50 * 100ms seconds.
```
Why the shift from a single `sleep` being interrupted by signal, btw?

daviddrysdale commented on procdesc.cc:
```diff
@@ -853,51 +874,80 @@ TEST_F(PipePdfork, WildcardWait) {
+    // Child: blocked inside read(). No SIGINT handler.
+    if (verbose) fprintf(stderr, "[%d] child about to read()\n", getpid_());
+    // Child wait for shutdown message. This should never be received, since
+    // SIGINT should termine the process.
```
s/termine/terminate/

daviddrysdale commented on procdesc.cc:
```diff
@@ -853,51 +874,80 @@ TEST_F(PipePdfork, WildcardWait) {
+    close(pipefds[0]);
+    if (verbose) fprintf(stderr, "[%d] child started\n", getpid_());
+    SEND_INT_MESSAGE(pipefds[1], MSG_CHILD_STARTED);
+    // Child: blocked inside read(). No SIGINT handler.
```
This comment could probably be combined with the one 2 lines below.

daviddrysdale reviewed (commented):
I've forgotten how these tests work, so apologies if some of my queries are lacking in understanding&hellip;

arichardson committed:
Remove sleep() from WildcardWait test

Instead use EXPECT_PID_ZOMBIE(). While changing this also remove the
stale FreeBSD TODO, this bug report has been fixed since 2015.

arichardson committed:
Remove sleep() calls from PassProcessDescriptor

This was the last remaining sleep-based synchronization in procdesc.cc,
so we can also remove the TODO comment.

arichardson committed:
Avoid using sleep() in Capmode.NewThread

This also improves the test so that all threads/child processes send back
a success/failure status to the main test body.

arichardson committed:
Avoid using sleep() in WithFiles.AllowedMiscSyscalls

Use a pipe as with the other tests.

arichardson committed:
Remove sleep() calls from the socket tests

I believe this is the last test that includes sleep-based synchronization.
The total runtime is now 4-5 seconds instead of almost 60 before the
changes to remove sleep().

arichardson commented:
@daviddrysdale I believe I've addressed all comments now and also addressed the remaining sleep() calls.
Thanks for the detailed review!

daviddrysdale commented on capmode.cc:
```diff
@@ -568,68 +576,131 @@ FORK_TEST_F(WithFiles, AllowedMiscSyscalls) {
   if (early_child == 0) {
+    if (verbose) fprintf(stderr, "  first child started\n");
+    EXPECT_OK(close(proc_pipe[0]));
     // Child: wait and then confirm this process is unaffect by capability mode in the parent.
```
nit: existing typo s/unaffect/unaffected

daviddrysdale commented on capmode.cc:
```diff
@@ -568,68 +576,131 @@ FORK_TEST_F(WithFiles, AllowedMiscSyscalls) {
   }
 
   EXPECT_OK(cap_enter());  // Enter capability mode.
+  // And the same for the process (which should not be affected by capability mode)
```
I wasn't sure what "the same" meant here. 

Could we add a comment here like:
```c
// At this point the current process has both a child process and a 
// child thread that were created before entering capability mode.
//  - The child process is unaffected by capability mode.
//  - The child thread is affected by capability mode.
```

daviddrysdale commented on capmode.cc:
```diff
@@ -568,68 +576,131 @@ FORK_TEST_F(WithFiles, AllowedMiscSyscalls) {
+  // The child processes/threads return HasFailure(), so we depend on no prior errors.
+  ASSERT_FALSE(testing::Test::HasFailure())
+              << "Cannot continue test with pre-existing failures.";
+  if (verbose) fprintf(stderr, "  starting second child (in capability mode)\n");
```
nit: maybe add a blank line and a comment here?
```c

// Now that we're in capability mode, if we create a second child process
// it will be affected by capability mode.
```

daviddrysdale commented on capmode.cc:
```diff
@@ -568,68 +576,131 @@ FORK_TEST_F(WithFiles, AllowedMiscSyscalls) {
+  // Wait for the second child to exit (should get a zero exit code message).
+  AWAIT_INT_MESSAGE(proc_pipe[0], 0);
 
   // Fire off a new thread.
```
Maybe `// Fire off a new (second) child thread, which is also affected by capability mode`?

daviddrysdale commented on procdesc.cc:
```diff
@@ -1031,13 +1075,15 @@ TEST_F(PipePdfork, PassProcessDescriptor) {
   if (verbose) fprintf(stderr, "[%d] send process descriptor %d on socket\n", getpid_(), pd_);
   int rc = sendmsg(sock_fds[1], &mh, 0);
   EXPECT_OK(rc);
+  // Wait until the child has receved the process descriptor.
```
nit: s/receved/received/

daviddrysdale reviewed (approved):
Just some comment suggestions (many of which are asking you to put in explanations that I should have added in the first place&hellip;)

arichardson commented:
Thanks, I've folded those suggestions into the two relevant commits.

daviddrysdale commented:
Thank you!
